### DISCUSSION TITLE
Display a form during cpu intensive task

### DISCUSSION THREAD

**Discussion by PlanetEarthSoftware** [picoe/Eto#2710] [Q&A]
Is there a way to force Eto to layout and render a Form during a ᴄᴘᴜ intensive task; like Invalidate() for a Drawable. 

Minimal example attached (65 lines total).

Thanks in advance for any assistance.

[EtoCpuIntensiveTest.zip](https://github.com/user-attachments/files/18032393/EtoCpuIntensiveTest.zip)

**Comment by cwensley** [comment]
To keep UI usable, any CPU intensive stuff needs to be done on separate threads.  Using async/await works really well for this, for example this is how I would recommend doing something like this:

```c#
    private async void MainForm_Shown(object sender, EventArgs e)
    {
        MessageForm messageForm = new("busy", "test message");
        messageForm.Closeable = false;
        messageForm.Minimizable = false;
        messageForm.Maximizable = false;
        messageForm.Show();

        await Task.Run(async () =>
        {
            await Task.Delay(10000);
            
            // do actual CPU intensive work here.
        });

        messageForm.Close();
    }
```

**Reply by ktwrd** [reply]
@cwensley @PlanetEarthSoftware 
Using threads is also a good solution, since that's what I use in all of my Eto.Forms and WinForms projects. But you've gotta do your own error handling, since exceptions thrown in the thread won't be passed through, and need to be handled inside of the thread.
```csharp
private void buttonRun_Click(object sender, EventArgs e)
{
	new System.Threading.Thread(async () =>
	{
		try
		{
			Eto.Forms.Application.Instance.Invoke(() =>
			{
				Content.Enabled = false;
				labelStatus.Text = "Processing task";
			});
			await YourIntensiveTask();
			Eto.Forms.Application.Instance.Invoke(() =>
			{
				Content.Enabled = true;
				labelStatus.Text = "Done!";
			});
		}
		catch (Exception ex)
		{
			// error handling
			Eto.Forms.Application.Instance.Invoke(() =>
			{
				Content.Enabled = true;
				labelStatus.Text = $"Failed to run task ({ex.Message})";
			});
		}
	}).Start();
}
```

**Comment by PlanetEarthSoftware** [comment]
I have found `Eto.Forms.Application.Instance.RunIteration();` which does enough of that that I require.

It would be nice if I could stop the menu and tool bars being active when `RunIteration()` runs, as if a modal dialog was open.
